feat(node): new observability architecture + events

karlem opened the pull request:
Close #1054

cryptoAtwill commented on fendermint/vm/topdown/src/observe.rs:
```diff
@@ -0,0 +1,240 @@
+#[derive(Debug)]
+pub struct ParentFinalityPeerVoteSent<'a> {
+    pub block_height: BlockHeight,
+    pub block_hash: BlockHashHex<'a>,
```
why keep these two fields if they are not used?

karlem replied:
They will be stored in the tracing event right.

cryptoAtwill commented on fendermint/vm/topdown/src/voting.rs:
```diff
@@ -288,14 +314,30 @@ where
+        parent_block_height: BlockHeight,
+        parent_block_hash: V,
+        proposer: Option<&str>,
+        local_block_height: Option<BlockHeight>,
```
what does this parameter mean? Like current fendermint block height?

karlem replied:
Yes exactly. I think it should be the Tendermint block. We can change the name if you find it confusing.

cryptoAtwill commented on fendermint/vm/topdown/src/voting.rs:
```diff
@@ -418,6 +460,14 @@ pub async fn publish_vote_loop<V, F>(
+
+                    let block_hash_hex = hex::encode(&next_hash);
+
+                    emit(ParentFinalityPeerVoteSent {
```
why not handle hex conversion in the method? It says a lot of `hex::encode` and one does not really have to import hex because of tracing

karlem replied:
Good point!

cryptoAtwill commented on fendermint/vm/topdown/src/proxy.rs:
```diff
@@ -64,23 +68,36 @@ impl IPCProviderProxy {
-        self.ipc_provider
-            .get_block_hash(&self.parent_subnet, height as ChainEpoch)
-            .await
+        emit_event_with_latency(
```
I kind of feel this is a bit intrusive? It's forcing the app to emit event and install the crates in monitor?

karlem replied:
I am not sure I follow. Could you please elaborate?

cryptoAtwill replied:
I mean this changes the implementation of measure latency by default. Now, if I composite the current struct, `emit_event_with_latency` is called, maybe I want to measure the whole process?

 In java or node, we can use decorators. Here, maybe in the caller we use sth like at the outest level:
```rust
emit_event_with_latency(finality.get_genesis_epoch(), ...).await
```
so that we can easily remove it.

karlem replied:
Yeah, I was considering to have another struct (implementing the same trait) that just wraps this one and measures the latency and emits the event. Would you be happier with that?

cryptoAtwill replied:
sounds good, my main concern is the measurement stays with this implementation but cannot easily remove it.

karlem commented:
@cryptoAtwill Alright, I have resolved all the comments.

raulk commented on fendermint/app/src/app.rs:
```diff
@@ -620,11 +623,42 @@ where
 
         let response = match result {
-            Err(e) => invalid_check_tx(AppError::InvalidEncoding, e.description),
+            Err(e) => {
```
Oh dear, these recursive results are such a telling symptom of just how unergonomic the nested interpreter design is /tech_debt 😢

karlem replied:
Yes indeed. I think it might be also worth to parse the transaction directly here - with help of external library.

raulk commented on fendermint/app/src/app.rs:
```diff
@@ -620,11 +623,42 @@ where
-                Ok(Err(InvalidSignature(d))) => invalid_check_tx(AppError::InvalidSignature, d),
-                Ok(Ok(ret)) => to_check_tx(ret),
+                Err(IllegalMessage) => {
+                    emit(MpoolReceivedInvalidMessage {
+                        reason: "IllegalMessage",
+                        description: "",
+                    });
```
1. What do you think about merging the `MpoolReceivedInvalidMessage` and `MpoolReceived` events into a single type (`MpoolReceived`), where `accept` determines whether it was accepted or not?
2. I would merge the reason and the description into a single field: `reason`, and [`format!()`](https://doc.rust-lang.org/std/macro.format.html) the original reason and description into it. You can then leverage the derived `Debug` implementation of the `AppError` enum to drop the inline strings (e.g. `"InvalidEncoding"`).
3. You can initialize a mutable struct at the top, update it throughout the nested `match`es, and emit it at the end, so the code is less pedantic.

karlem replied:
I wanted to follow your suggestion, but that would require making a lot of fields in MpoolReceived optional. This is particularly necessary for cases like InvalidEncoding, where the transaction fails to parse the raw bytes. However, I suppose we can live with the optional types.

karlem replied:
Alright, I have implemented it using options.

raulk commented on fendermint/app/src/app.rs:
```diff
@@ -650,7 +684,13 @@ where
+
+        emit(BlockProposalSent {
+            height: request.height.value(),
+            tx_count: txs.len(),
```
It would be useful to print our own validator address here, which is part of `request::PrepareProposal`.

karlem replied:
Yes, good point. Added it.

raulk commented on fendermint/app/src/app.rs:
```diff
@@ -666,26 +706,40 @@ where
+
+        emit(BlockProposalReceived {
+            height: request.height.value(),
+            hash: HexEncodableBlockHash(request.hash.into()),
```
Looks like `HexEncodableBlockHash` type is repeated in two crates. We could move somewhere more common (e.g. `ipc-observability`), under a `serde` module.

karlem replied:
Agreed, done.

raulk commented on fendermint/app/src/app.rs:
```diff
@@ -666,26 +706,40 @@ where
+            hash: HexEncodableBlockHash(request.hash.into()),
+            size: size_txs,
+            tx_count: num_txs,
+            validator: request.proposer_address.to_string().as_str(),
```
The `account::Id` type implements `fmt::Display`, so it should be fine to use the type as-is.

raulk commented on fendermint/app/src/app.rs:
```diff
@@ -666,26 +706,40 @@ where
-            Ok(response::ProcessProposal::Reject)
+        match process_result {
+            Ok(_) => {
+                emit(BlockProposalAccepted {
+                    height: request.height.value(),
+                    hash: HexEncodableBlockHash(request.hash.into()),
+                    size: size_txs,
+                    tx_count: num_txs,
+                    validator: request.proposer_address.to_string().as_str(),
+                });
+                Ok(response::ProcessProposal::Accept)
+            }
+            Err(e) => {
+                emit(BlockProposalRejected {
+                    height: request.height.value(),
+                    size: size_txs,
+                    tx_count: num_txs,
+                    validator: request.proposer_address.to_string().as_str(),
+                    reason: e.to_string().as_str(),
+                });
```
Similar to the above, we could merge these types into a single `BlockProposalEvaluated` type.

karlem replied:
Yes. I have changed it to be more DRY friendly!

raulk reviewed (commented):
1/3 reviewed.

karlem commented on fendermint/app/src/app.rs:
```diff
@@ -620,11 +623,42 @@ where
+                    invalid_check_tx(AppError::InvalidSignature, d)
+                }
+                Ok(Ok(ret)) => {
+                    emit(MpoolReceived {
+                        from: ret.message.from.to_string().as_str(),
+                        to: ret.message.to.to_string().as_str(),
+                        value: ret.message.value.to_string().as_str(),
+                        param_len: ret.message.params.len(),
+                        gas_limit: ret.message.gas_limit,
+                        fee_cap: ret.message.gas_fee_cap.to_string().as_str(),
+                        premium: ret.message.gas_premium.to_string().as_str(),
+                        accept: ret.exit_code.is_success(),
```
It does indeed. I needed some of the parameters to be strings (for instance, for metrics labels) and wanted to have generic events. But you are right; it would be more efficient to just use the types directly (the ones that implement Display).

raulk replied:
Got it, can we attempt to preserve the types during emit, and push the conversion deeper in, to where it's actually needed (emitting the metric with labels)?

karlem replied:
Yes, I have already done it :)

karlem commented on fendermint/app/src/app.rs:
```diff
@@ -620,11 +623,42 @@ where
+                        gas_limit: ret.message.gas_limit,
+                        fee_cap: ret.message.gas_fee_cap.to_string().as_str(),
+                        premium: ret.message.gas_premium.to_string().as_str(),
+                        accept: ret.exit_code.is_success(),
```
Yeah, that's a very good point! It will also simplify the code give the suggestion about merging the events together.

karlem commented on Cargo.toml:
```diff
@@ -168,6 +173,7 @@ ipc-provider = { path = "ipc/provider" }
 ipc-wallet = { path = "ipc/wallet", features = ["with-ethers"] }
 ipc_ipld_resolver = { path = "ipld/resolver" }
 ipc-types = { path = "ipc/types" }
+ipc-observability = { path = "ipc/observability" }
 ipc_actors_abis = { path = "contracts/binding" }
```
Created an [issue](https://github.com/consensus-shipyard/ipc/issues/1076)

karlem commented on fendermint/app/options/src/lib.rs:
```diff
@@ -102,13 +103,44 @@ pub struct Options {
     #[arg(long, env = "FM_CONFIG_DIR")]
     config_dir: Option<PathBuf>,
 
+    // TODO Karel - move all FM_LOG_FILE* flags to a configuration file instead
```
Alright - there is an [issue](https://github.com/consensus-shipyard/ipc/issues/1071) for it. I can pick it up as long as we finish review of this PR.

karlem replied:
https://github.com/consensus-shipyard/ipc/pull/1078 dropping it here to have the configuration PR linked. It's almost done!

raulk commented on fendermint/vm/interpreter/src/fvm/state/exec.rs:
```diff
@@ -350,3 +351,39 @@ fn check_error(e: anyhow::Error) -> (ApplyRet, ActorAddressMap) {
     (ret, Default::default())
 }
+
+pub struct ElapsedExecution<'a, DB>
+where
+    DB: Blockstore + Clone + 'static,
+{
+    state: &'a mut FvmExecState<DB>,
+}
+
+impl<'a, DB> ElapsedExecution<'a, DB>
+where
+    DB: Blockstore + Clone + 'static,
+{
+    pub fn new(state: &'a mut FvmExecState<DB>) -> Self {
+        Self { state }
+    }
+
+    pub fn execute_implicit(
+        &mut self,
+        msg: Message,
+    ) -> anyhow::Result<(ApplyRet, ActorAddressMap, Duration)> {
+        let start = Instant::now();
+        let result = self.state.execute_implicit(msg)?;
+        let duration = start.elapsed();
+        Ok((result.0, result.1, duration))
+    }
+
+    pub fn execute_explicit(
+        &mut self,
+        msg: Message,
+    ) -> anyhow::Result<(ApplyRet, ActorAddressMap, Duration)> {
+        let start = Instant::now();
+        let result = self.state.execute_explicit(msg)?;
+        let duration = start.elapsed();
+        Ok((result.0, result.1, duration))
+    }
+}
```
Introducing a new type seems a bit heavy-handed to me. It's only used in 2 sites, and I think it's acceptable to add a helper function to the observability crate -- this would allow us to time any closure.

```rust
fn measure_time<F, T>(f: F) -> (T, std::time::Duration)
where
    F: FnOnce() -> T,
{
    let start = Instant::now();
    let result = f();
    let duration = start.elapsed();
    (result, duration)
}
```

karlem replied:
Yeah I agree. This change was a reaction to this [comment here](https://github.com/consensus-shipyard/ipc/pull/1064#discussion_r1679114289)

raulk reviewed (commented):
2/3.

raulk commented on fendermint/vm/interpreter/src/bytes.rs:
```diff
@@ -125,13 +126,17 @@ where
                 "rejecting block: too many messages"
             );
-            return Ok(false);
+            return Err(ProcessError::TooManyMessages(msgs.len()));
```
This changes the return semantics of this method. In fact, this appears to be a breaking change. The caller (`App::process_proposal()`) would previously return a successful rejection to CometBFT, it now propagates the error.

I don't think the motivation (to capture a reason in the event) justifies the risk of a functional regression in this part of the codebase. Let's revert these changes and settle for not capturing the reason for the rejection in the trace.

karlem replied:
Yes I guess that makes sense. To me though I did not see why would we want to have it succeed in the case. Also I have changed the error handling as well. But I understand why the risk on correct migration is higher then benefit of reason.

raulk reviewed (approved):
Overall, great job here! This will bring about a massive improvement in observability and monitoring. Excited to see it operating live. Kudos on your first major epic, @karlem! 👏 

If you're able to get the changes addressed in the next hours, we should be able to merge today and release tomorrow with Axon r02! 🚀

karlem commented on fendermint/app/options/src/lib.rs:
```diff
@@ -162,6 +194,17 @@ impl Options {
         }
     }
 
+    pub fn log_file_config(&self) -> FileLayerConfig {
+        FileLayerConfig {
+            enabled: self.log_file_enabled.unwrap_or(false),
+            directory: self.log_dir.clone(),
+            max_log_files: self.max_log_files,
+            rotation: self.log_files_rotation.clone(),
+            domain_filter: self.domains_filter.clone(),
+            events_filter: self.events_filter.clone(),
+        }
+    }
```
[Addressed here](https://github.com/consensus-shipyard/ipc/pull/1078)

karlem commented on fendermint/vm/topdown/src/voting.rs:
```diff
@@ -278,6 +294,14 @@ where
+                    block_height: *block_height,
+                    block_hash: HexEncodableBlockHash(block_hash.as_ref().to_vec()),
+                    // TODO Karel - just placeholder - need to use real commitment once implemented
+                    commitment_hash: HexEncodableBlockHash(block_hash.as_ref().to_vec()),
```
Yes, good idea!

karlem commented:
@raulk Thanks! I have addressed all your comments, and it should be ready to go. However, please bear in mind that there are still two outstanding PRs that require review.

raulk commented on fendermint/app/src/observe.rs:
```diff
@@ -113,25 +114,48 @@ impl Recordable for BlockCommitted {
 
 impl Recordable for MpoolReceived {
     fn record_metrics(&self) {
+        let from = self
+            .message
+            .as_ref()
+            .map(|m| m.from.to_string())
+            .unwrap_or("".to_string());
+
```
Let's drop the `from` label from the metric. It's unbounded cardinality, and Prometheus cannot deal with that. Read more here: https://prometheus.io/docs/practices/naming/#labels

karlem replied:
Good point.

raulk commented on fendermint/vm/interpreter/src/fvm/observe.rs:
```diff
@@ -18,63 +20,40 @@ register_metrics! {
+            MsgExecPurpose::Call => EXEC_FVM_CALL_EXECUTION_TIME_SECS.observe(self.duration),
+        }
     }
 }
```
Nice! It's all coming together. This would've been excrutiatingly difficult to do with the previous code.

karlem replied:
Yes, it is also much cleaner this way.

raulk commented on fendermint/vm/topdown/src/observe.rs:
```diff
@@ -214,27 +214,27 @@ mod tests {
             parent_height: 0,
             block_hash: HexEncodableBlockHash(hash.clone()),
             local_height: Some(0),
-            proposer: Some("proposer"),
```
What's this here? Is this a TODO?

raulk replied:
Ah, it's just a test. Ignore my comment ;-)

raulk commented on fendermint/vm/interpreter/src/fvm/state/exec.rs:
```diff
@@ -177,11 +176,13 @@ where
 
     /// Execute message implicitly.
     pub fn execute_implicit(&mut self, msg: Message) -> ExecResult {
+        // TODO Karel - measure here
```
Do we _really_? I think it's fine not to overload this component with instrumentation, and just leave it up to the user -- just like we're doing now.

karlem replied:
Not really yeah :) I forgot to remove my comment. It should not be there..

raulk reviewed (commented):
A few follow-up comments. I'm scrutinizing the ProcessResult and ProcessError changes as we speak, so expect one more final review of that in the next minutes.

raulk commented on fendermint/app/src/app.rs:
```diff
@@ -666,27 +688,43 @@ where
         let num_txs = txs.len();
 
-        let accept = self
+        let process_result = self
             .interpreter
             .process(self.chain_env.clone(), txs)
-            .await
-            .context("failed to process proposal")?;
```
Dug into this further:

- Returning an error from the `Application` will induce a panic and cause the Tower Service to crash. 
- This approach is sensible if (a) Tower is configured to restart the service upon crash, AND (b) we're in the middle of some state update (e.g. after begin_block and before end_block). In that case, we want to reset transient state and start over.
- However, (b) is not true for process_proposal -- this is a stateless method. So returning an error here is unnecessary and problematic.

Conclusion:

- Thanks to your diff, we discovered a deeper issue! 🌟 
- But because this is a delicate part of the system, I prefer to extract this diff out to its own changeset and changelog entry.
- Behaviourally speaking, these changes seem reasonable, but I have some hesitations on introducing new types to signal a binary outcome (idiomatic way is to use Result or Option, either flattened or nested).

Next steps:

- Let's back out these changes and forego setting a reason for now.
- Open a separate PR with your changes that also prevents `process_proposal` from propagating an error.

raulk reviewed (approved):
Feel free to merge once the above comments have been addressed ;-)
